File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
# 프로그래머스 87694 아이템 줍기 LV.3
from collections import deque


def solution(rectangle, characterX, characterY, itemX, itemY):
    field = [[-1] * 102 for _ in range(102)]

    for r in rectangle:
        x1, y1, x2, y2 = map(lambda x: x * 2, r)
        for i in range(x1, x2 + 1):
            for j in range(y1, y2 + 1):
                if x1 < i < x2 and y1 < j < y2:
                    field[i][j] = 0
                elif field[i][j] != 0:
                    field[i][j] = 1

    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    visited = [[False] * 102 for _ in range(102)]
    visited[characterX * 2][characterY * 2] = True

    def bfs(x, y):
        q = deque()
        q.append([x, y, 0])
        while q:
            x, y, step = q.popleft()

            if x == itemX * 2 and y == itemY * 2:
                return step // 2

            for i in range(4):
                nx, ny = x + dx[i], y + dy[i]
                if field[nx][ny] == -1:
                    continue
                if not visited[nx][ny] and field[nx][ny] == 1:
                    visited[nx][ny] = True
                    q.append([nx, ny, step + 1])

    return bfs(characterX * 2, characterY * 2)

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.java
import java.util.*;

class Solution {
    int[] dy = new int[]{-1, 0, 1, 0}, dx = new int[]{0, 1, 0, -1};
    
    public int[][] convertRectangle(int[][] rectangle) {
        int[][] field = new int[101][101];
        
        for (int[] rect : rectangle) {
            int leftX = rect[1] * 2, leftY = rect[0] * 2, rightX = rect[3] * 2, rightY = rect[2] * 2;
            
            for (int i = leftY; i <= rightY; i++) {
                for (int j = leftX; j <= rightX; j++) {
                    if (i == leftY || i == rightY || j == leftX || j == rightX) {
                        if (field[i][j] != 2) {
                            field[i][j] = 1;
                        }
                    } else {
                        field[i][j] = 2;
                    }
                }
            }
        }
        
        return field;
    }
    
    public int bfs(int[][] field, int startX, int startY, int endX, int endY) {
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{startY, startX, 0});
        boolean[][] visited = new boolean[101][101];
        visited[startY][startX] = true;
        
        while (!q.isEmpty()) {
            int[] coord = q.poll();
            int y = coord[0], x = coord[1], dist = coord[2];
            
            if (y == endY && x == endX) return dist / 2;
            
            for (int i = 0; i < 4; i++) {
                int ny = y + dy[i], nx = x + dx[i];
                
                if (checkRange(ny, nx) && !visited[ny][nx] && checkEdge(field, ny, nx)) {
                    q.offer(new int[]{ny, nx, dist + 1});
                    visited[ny][nx] = true;
                }
            }
        }
        
        return 0;
    }
    
    public boolean checkRange(int nextY, int nextX) {
        return 1 <= nextY && nextY < 101 && 1 <= nextX && nextX < 101;
    }
    
    public boolean checkEdge(int[][] field, int nextY, int nextX) {
        return field[nextY][nextX] == 1;
    }
    
    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
        int startX = characterY, startY = characterX, endX = itemY, endY = itemX;
        
        int[][] field = convertRectangle(rectangle);
        
        return bfs(field, startX * 2, startY * 2, endX * 2, endY * 2);
    }
}